Use whiptail instead of basic read in the repository setup script

sdumitriu opened the pull request.

sdumitriu committed:
Use whiptail instead of basic read in the repository setup script
Check early on if the locally build CARDS image is a production one
+ wording changes

sdumitriu committed:
[misc] Add license header to the setup script

sdumitriu committed:
Also include the local snapshot version in the list of versions to choose from

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,87 @@
+#  limitations under the License.
+
+# Check the user's authorship configuration
+if ! whiptail --backtitle "New CARDS repository setup" --title "Please confirm git authorship details" --yesno "Committing as $(git config user.name) <$(git config user.email)>, OK?" 8 78
```
If the git `user.name` and `user.email` properties have not been set, the message shown will be `Committing as  <>, OK?` and the user will still have the option of selecting `<Yes>` which will ultimately lead to a failed `git commit`. Therefore, I think before this `whiptail --yesno` is even called, the script should verify that both the `user.name` and `user.email` git properties have been set.

IntegralProgrammer commented on get_cards_platform_jars.sh:
```diff
@@ -36,6 +36,19 @@ do
+      VARIANT=0
+    fi
   fi
 done
```
Should this be changed to use `whiptail` as well?

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,87 @@
 # Get the latest CARDS version from Nexus
-CARDS_VERSION=$(curl -s https://nexus.phenotips.org/nexus/content/repositories/releases/io/uhndata/cards/cards-modules/maven-metadata.xml | grep --max-count=1 '<release>' | cut '-d>' -f2 | cut '-d<' -f1)
-if [[ -z $CARDS_VERSION ]]
+declare -a RELEASED_VERSIONS=( $(curl -s https://nexus.phenotips.org/nexus/content/repositories/releases/io/uhndata/cards/cards-modules/maven-metadata.xml | grep '<version>' | tac | cut '-d>' -f2 | cut '-d<' -f1) )
+declare -i RELEASED_VERSIONS_COUNT=${#RELEASED_VERSIONS[*]}
+# Look on the local disk
+LOCAL_CARDS_VERSION=$(grep --max-count=1 '^  <version>' ../cards/pom.xml 2>/dev/null | cut '-d>' -f2 | cut '-d<' -f1)
+declare VERSION
+if [[ -n $LOCAL_CARDS_VERSION ]]
+then
+  VERSION="${LOCAL_CARDS_VERSION} ${LOCAL_CARDS_VERSION} OFF "
+fi
+for (( i=0 ; i<RELEASED_VERSIONS_COUNT; i++ ))
+do
+  VERSION="${VERSION} ${RELEASED_VERSIONS[$i]} ${RELEASED_VERSIONS[$i]}"
+  if [[ $i == 0 ]]
+  then
+    VERSION+=" ON "
+  else
+    VERSION+=" OFF "
+  fi
+done
+CARDS_VERSION=$(whiptail --backtitle "New CARDS repository setup" --title "Base CARDS version" --radiolist --notags "Which version of the CARDS platform should be used?" 38 78 30 $VERSION 3>&1 1>&2 2>&3)
```
If the call to `curl -s https://nexus.phenotips.org/nexus/content/repositories/releases/io/uhndata/cards/cards-modules/maven-metadata.xml` fails and thus the only listed version is the one available locally, it will, by default, be unselected and therefore it is possible to proceed without actually setting the `CARDS_VERSION` variable.

Also, I don't like the assumption that the main CARDS repository would be located locally at `../cards`. I think instead we should ask for this path at the beginning of the script.

sdumitriu replied:
I don't want to make this too complicated, getting these values is a best-effort task. Instead I think we should have an `Other` option which will bring up an input box where a custom version can be entered.

IntegralProgrammer commented on setup.sh:
```diff
@@ -87,46 +115,19 @@ fi
 declare -a features=( $(find ~/.m2/repository/io/uhndata/cards/cards/${CARDS_VERSION}/ -type f -name '*slingosgifeature' | sed -r -e "s/.*${CARDS_VERSION}-(.*).slingosgifeature/-e \1/" -e /cards/d) )
 features=( $(find ~/.m2/repository/io/uhndata/cards/ -type f -name "*${CARDS_VERSION}.slingosgifeature" | grep -v -e 'cards4' | grep -v ${features[*]} | sed -r -e "s/.*\/(.*)-${CARDS_VERSION}.slingosgifeature/\1/") )
```
The problem with this is that it only searches for available CARDS features under the local `~/.m2/repository` directory which none will be found there unless CARDS has already been built locally. Do we have available anywhere (for example, on GitHub or `nexus.phenotips.org`) lists of available modules for each released version of CARDS? I think those should be used if a released version of CARDS is selected and only if a _SNAPSHOT_ version of CARDS is selected should we look under `~/.m2/repository` for available features.

sdumitriu replied:
Good idea.

sdumitriu replied:
Now we're looking in the published docker images.

IntegralProgrammer reviewed (changes requested):
In addition to these comments, I cannot build the repository (containing my test project) after running `./setup.sh` as:

- `cards-parent-1.pom` cannot be found (see: https://github.com/data-team-uhn/cards-project-template/commit/f7f7c8322a7959ab4a28eaae3c27c74c8df537f3#r135277441)
- After manually replacing `1` with `CARDS-0.9.22-SNAPSHOT`, the build still fails at the RAT license check as `docker/project_code.txt` and `docker/project_name.txt` are not being properly excluded from the RAT check (see: https://github.com/data-team-uhn/cards-parent/commit/9498571743fdcbcfa71535c4e7ab9c55b5215470#r135276914, https://github.com/data-team-uhn/cards-parent/commit/9498571743fdcbcfa71535c4e7ab9c55b5215470#r135276939)

sdumitriu committed:
[misc] We're not using a separate cards-parent anymore

sdumitriu committed:
[misc] Abort if there's no git committer set up

sdumitriu committed:
[misc] Use maven central instead of phenotips nexus to look for released versions.
[misc] Allow entering a custom version
[misc] Always have a radio button selected when choosing the version

sdumitriu committed:
[misc] Mark the generic cards dependencies repo as ignored by git

sdumitriu committed:
[misc] Add SCM info to Maven, since by default it is inherited from the parent pom

sdumitriu committed:
[misc] Update the maven-release-plugin configuration to avoid cards-specific work

sdumitriu committed:
[misc] Use the base docker image to look for extra features to enable instead of the local .m2 repository

sdumitriu committed:
[misc] Maven property values are computed in each project, so we must hardcode the cards.version property

sdumitriu committed:
[misc] Add instructions for quickly running a transient docker image

sdumitriu committed:
[misc] Update the location of the docker_compose_env.json file in the documentation

sdumitriu committed:
[misc] Use whiptail in the get_cards_platform_jars script

IntegralProgrammer commented:
Setting CARDS data access permissions with the `PERMISSIONS` environment variable is broken. For example, I can use https://github.com/data-team-uhn/cards-deploy-tool as `python3 generate_compose_yaml.py --cards_docker_image ghcr.io/data-team-uhn/cards:0.9.23 --oak_filesystem --composum`, edit the generated `docker-compose.yml` file to specify the `PERMISSIONS=trusted` environment variable for the `cardsinitial` container and CARDS will start with the _TrustedUsers_ group created. If I do the same process but using an image generated with `./setup.sh` (eg. `python3 generate_compose_yaml.py --cards_docker_image cards/cards4test:latest --oak_filesystem --composum`) the _TrustedUsers_ group will not be created.

The root cause of this issue is that the `/sling-features.json` file is not updated with the new project and therefore https://github.com/data-team-uhn/cards/blob/3f0d0b656a83fbb73893c1324f0941b06be7b694/distribution/docker_entry.sh#L96 only ever populates the `PROJECT_REQUIRED_FEATURES` variable with an empty string, thus, the specification to use the `cards-dataentry` module in `permissions_trusted` mode never actually happens.

So therefore:

- The `/sling-features.json` file in the built container should specify *all* required modules for the project.
- Given that these features would then be specified in the `/sling-features.json` file, they should no longer be specified in the `ADDITIONAL_SLING_FEATURES` environment variable - this makes for neater code and documentation anyways.
- Given that the default permissions are being set in the `/sling-features.json` file that is baked into the project Docker image, there should be a whiptail prompt generated by `setup.sh` which asks for the default permissions scheme to use (`open`, `trusted`, `ownership`).

sdumitriu committed:
[misc] Remove debug echo

sdumitriu committed:
[misc] Use whiptail for output messages in the get_cards_platform_jars script

sdumitriu committed:
Declare feature dependencies once in the root pom
Generate the files needed for start_cards.sh and docker_entry.sh at build time based on these dependencies

sdumitriu committed:
Don't output a success message when getting the platform jars with a specific version

sdumitriu committed:
Ask about the permission scheme to use during repo setup

sdumitriu committed:
Only generate the sling-features.json file, no need for a separate flat .dependencies file

IntegralProgrammer commented on setup.sh:
```diff
@@ -58,90 +115,80 @@ else
-until [[ $HAS_BACKEND == "yes" || $HAS_BACKEND == "no" ]]
-do
-  read -e -p "[yes|no] " -i "no" HAS_BACKEND
+./get_cards_platform_jars.sh ${CARDS_VERSION}
```
Exit here with an error if `get_cards_platform_jars.sh` fails.

```suggestion
./get_cards_platform_jars.sh ${CARDS_VERSION} || exit -1
```

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,102 @@
 
-echo "What is the project's codename, e.g. cards4sparc, cards4heracles"
-read -e -i cards4 PROJECT_CODENAME
+PROJECT_CODENAME=$(whiptail --backtitle "New CARDS repository setup" --title "Project name" --inputbox "What is the project's codename? e.g. cards4sparc, cards4lfs" 8 78 cards4 3>&1 1>&2 2>&3)
```
Docker image names cannot contain any upper-case characters therefore this should be specified and verified here.

sdumitriu replied:
Done.

IntegralProgrammer reviewed (changes requested):
In addition to my suggested improvements, I believe that the following should be added as well:

- An error should be thrown if `mvn clean install -P production,latest` is executed but the `cards/cards:latest` image is not a production image.
- The problem with `docker/pom.xml` (line 133)
```xml
<configuration>  
  <executable>bash</executable>
  <workingDirectory>.</workingDirectory>
  <commandlineArgs>-c "mkdir -p .m2/repository"</commandlineArgs>
</configuration>
```
is that (because of the `-p` flag) it will not fail if `.m2/repository` already exists so if the `.m2/repository` directory exists and has already been populated with JARs from a prior execution of `mvn clean install -P production...` those JARs will be copied into the "development" image even though such a thing should only happen in _production_ images.

- If, within `setup.sh`, we choose to use a `-SNAPSHOT` version as the CARDS base and the local `cards/cards:latest` image is a _development_ image, it will fail with _"Error: Failed to extract the needed files from the generic CARDS image."_ (as expected) but will then continue with the rest of the `setup.sh` process (see my code suggestion at line 125 in `setup.sh`) although in a broken state as the _"Which permission scheme should be used?"_ prompt will be blank. So, as I see it, `-SNAPSHOT` versions of CARDS should only be listed if the `cards/cards:latest` local image is a _production_ image. This would still allow developers to build _development_ images of their CARDS project (as the appropriate JAR directories would be mounted to the project container) it just means that a _production_ `-SNAPSHOT` container image would need to first be built so that `setup.sh` can be run and the project could be properly configured.

- If, within `setup.sh`, we choose to use a `-SNAPSHOT` version as the CARDS base and the local `cards/cards:latest` image is a _production_ image, the _"Which permission scheme should be used?"_ whiptail dialog will list permission labels whose lengths exceed the width of the dialog box instead of the expected `ownership`, `open`, and `trusted` labels.

![screenshot_2024_04_29-16-25-02](https://github.com/data-team-uhn/cards-project-template/assets/5349728/3edcc21c-2ed9-488d-9971-ecca6203ad8a)

sdumitriu committed:
[misc] If the platform jars cannot be extracted from the docker image, abort the repo setup procedure

Co-authored-by: Michael Lescisin <IntegralProgrammer@users.noreply.github.com>

sdumitriu committed:
Project codename should be all lowercase

sdumitriu committed:
Clean up any previous production jars when building a development docker image

sdumitriu committed:
Fail the build if a production image was requested but the base cards image isn't a production one

sdumitriu committed:
Wrong detection of the permission schemes when building on top of a local snapshot production image

sdumitriu committed:
Wrong detection of extra features when running the setup script on top of a snapshot cards build

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,109 @@
+declare -a RELEASED_VERSIONS=( $(curl -s https://repo1.maven.org/maven2/io/uhndata/cards/cards-modules/maven-metadata.xml | grep '<version>' | tac | cut '-d>' -f2 | cut '-d<' -f1) )
+declare -i RELEASED_VERSIONS_COUNT=${#RELEASED_VERSIONS[*]}
+# Look on the local disk
+LOCAL_CARDS_VERSION=$(grep --max-count=1 '^  <version>' ../cards/pom.xml 2>/dev/null | cut '-d>' -f2 | cut '-d<' -f1)
```
I don't like the assumption that the CARDS base repository will always be available at `../cards` if it exists locally. Perhaps it might be a better idea to list locally-available CARDS versions by looking under `~/.m2/repository/io/uhndata/cards/cards`?

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,109 @@
+LOCAL_CARDS_VERSION=$(grep --max-count=1 '^  <version>' ../cards/pom.xml 2>/dev/null | cut '-d>' -f2 | cut '-d<' -f1)
+declare VERSION
+declare HAS_DEFAULT
+for (( i=0 ; i<RELEASED_VERSIONS_COUNT; i++ ))
+do
+  VERSION="${VERSION} ${RELEASED_VERSIONS[$i]} ${RELEASED_VERSIONS[$i]} ${HAS_DEFAULT:-ON} "
+  HAS_DEFAULT=OFF
+done
+if [[ -n $LOCAL_CARDS_VERSION ]]
+then
+  VERSION="${LOCAL_CARDS_VERSION} ${LOCAL_CARDS_VERSION} ${HAS_DEFAULT:-ON} ${VERSION}"
+  HAS_DEFAULT=OFF
+fi
```
I think this should be amended so that `$LOCAL_CARDS_VERSION` is not listed if it is already included in the `$RELEASED_VERSIONS` list. For example, if the local version of CARDS is `0.9.22`, `0.9.22` would be listed twice as it is available both locally and in the list of released versions obtained from the Maven repository.

But given my above comment regarding switching from checking `../cards/pom.xml` for the locally available CARDS version to looking under `~/.m2/repository/io/uhndata/cards/cards` we could end up having more than one locally available version. Therefore, what would be best would be to list all locally available versions as well as the versions listed on the Maven repository but ensuring that no single version gets listed more than once.

sdumitriu commented:
@IntegralProgrammer What if instead of looking at local snapshots we just look for a `cards/cards:latest` docker image and get the version from inside it? We will require that image anyway.

sdumitriu committed:
Extract the snapshot version from the latest cards image instead of the local repository

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,109 @@
+declare -a RELEASED_VERSIONS=( $(curl -s https://repo1.maven.org/maven2/io/uhndata/cards/cards-modules/maven-metadata.xml | grep '<version>' | tac | cut '-d>' -f2 | cut '-d<' -f1) )
+declare -i RELEASED_VERSIONS_COUNT=${#RELEASED_VERSIONS[*]}
+# Look on the local disk
+LOCAL_CARDS_VERSION=$(docker run --rm  --entrypoint /bin/sh cards/cards:latest -c "ls /root/.m2/repository/io/uhndata/cards/cards-parent/ | grep SNAPSHOT")
```
I think it would be best to first check that `cards/cards:latest` is available locally and only run this command if such is the case. Otherwise, this command will attempt to pull `cards/cards:latest` from Docker Hub if no `cards/cards:latest` image exists locally. Please see https://github.com/data-team-uhn/cards/blob/dev/Utilities/Packaging/Docker/build_self_contained.sh#L28-L34 for an example of how we do this check.

sdumitriu replied:
Done.

sdumitriu committed:
Before trying to get the snapshot version from the local cards image, first check that such an image actually exists to avoid trying to fetch from dockerhub

sdumitriu committed:
Before trying to get the snapshot version from the local cards image, first check that such an image actually exists to avoid trying to fetch from dockerhub

sdumitriu committed:
Silence the error output when local cards image is not a production image

IntegralProgrammer commented on docker/check_production_base_image.sh:
```diff
@@ -0,0 +1,30 @@
+if [[ $STATUS -eq 125 ]]
+then
+  echo "No local image found, please rebuild it following the instructions for building a production-ready docker image in the CARDS README, then try again." 1>&2
+  exit -1
```
We should do the check for the presence of the local image in the same way that we are doing that check in https://github.com/data-team-uhn/cards-project-template/blob/08d64fbbcee06486ac34c3945c177f7f0f2b4188/setup.sh#L38-L39

sdumitriu replied:
Done.

IntegralProgrammer commented on get_cards_platform_jars.sh:
```diff
@@ -18,34 +18,42 @@
+    if [[ $STATUS -eq 125 ]]
+    then
+      whiptail --msgbox "No local image found, please rebuild it following the instructions for building a production-ready docker image in the CARDS README, then try again." 9 78
+      exit -1
```
We should do the check for the presence of the local image in the same way that we are doing that check in https://github.com/data-team-uhn/cards-project-template/blob/08d64fbbcee06486ac34c3945c177f7f0f2b4188/setup.sh#L38-L39

sdumitriu replied:
Done.

sdumitriu committed:
Before trying to get the snapshot version from the local cards image, first check that such an image actually exists to avoid trying to fetch from dockerhub

sdumitriu committed:
Fix wrong phase used for building docker images

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,110 @@
 
 # Remove the cards4 prefix to get the short unprefixed name
 PROJECT_SHORTNAME=${PROJECT_CODENAME#cards4}
+if [[ $? == 1 ]]
+then
+  exit 1
+fi
```
What is the purpose of this?

sdumitriu replied:
That's the status code for choosing "Cancel". I consider that pressing Cancel in a dialog shout terminate the script. A pending improvement for the next PR is to gather all the data first, then perform all the chances, so that pushing Cancel at a later stage doesn't leave the repo in an intermediary state.

IntegralProgrammer replied:
Right, but isn't that already handled by lines 76-78

```bash
while [[ -z $PROJECT_CODENAME ]]
do
  PROJECT_CODENAME=$(whiptail --backtitle "New CARDS repository setup" --title "Project name" --inputbox "What is the project's codename? No uppercase letters are allowed, e.g. cards4sparc, cards4lfs" 8 78 cards4 3>&1 1>&2 2>&3)
  if [[ $? == 1 ]]
  then
    exit 1
  elif [[ ${PROJECT_CODENAME} != ${PROJECT_CODENAME,,} ]]
  then
    PROJECT_CODENAME=
  fi
done
```

sdumitriu replied:
You're right, it is.

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,110 @@
 do
-  echo "Please provide a logo for the project. It should be a file about 200px wide and 80px tall, and display well on a dark background."
-  read -e LOGO
+  LOGO=$(whiptail --backtitle "New CARDS repository setup" --title "Project logo" --inputbox "Please provide a logo for the project. It should be a file about 200px wide and 80px tall, and display well on a dark background." 8 78 3>&1 1>&2 2>&3)
```
```suggestion
  LOGO=$(whiptail --backtitle "New CARDS repository setup" --title "Project logo" --inputbox "Please provide a logo for the project. It should be a file about 200px wide and 80px tall, and display well on a dark background." 8 78 3>&1 1>&2 2>&3)
if [[ $? == 1 ]]
then
  exit 1
fi
```

IntegralProgrammer commented on setup.sh:
```diff
@@ -1,45 +1,110 @@
-echo
-echo "Please provide a logo to be displayed on a light background. If the same image as before can be used, just press enter."
-read -e LOGO_LIGHT
+LOGO_LIGHT=$(whiptail --backtitle "New CARDS repository setup" --title "Project logo" --inputbox "Please provide a logo to be displayed on a light background. If the same image as before can be used, just press enter." 8 78 3>&1 1>&2 2>&3)
```
```suggestion
LOGO_LIGHT=$(whiptail --backtitle "New CARDS repository setup" --title "Project logo" --inputbox "Please provide a logo to be displayed on a light background. If the same image as before can be used, just press enter." 8 78 3>&1 1>&2 2>&3)
if [[ $? == 1 ]]
then
  exit 1
fi
```

IntegralProgrammer commented on setup.sh:
```diff
@@ -58,90 +123,80 @@ else
-  fi
+  featurelist+="${features[$i]} ${features[$i]} OFF "
 done
+selectedFeatures=$(whiptail --backtitle "New CARDS repository setup" --title "Modules setup" --checklist --notags "Other features to enable?" 38 78 30 $featurelist 3>&1 1>&2 2>&3)
```
```suggestion
selectedFeatures=$(whiptail --backtitle "New CARDS repository setup" --title "Modules setup" --checklist --notags "Other features to enable?" 38 78 30 $featurelist 3>&1 1>&2 2>&3)
if [[ $? == 1 ]]
then
  exit 1
fi
```

IntegralProgrammer commented on docker/check_production_base_image.sh:
```diff
@@ -0,0 +1,37 @@
+    exit -1
+  fi
+fi
+
```
I think it would also be good to check that the local `cards/cards:latest` image is for the _SNAPSHOT_ version of CARDS specified by the specific project. For example, if I run `setup.sh` and specify `0.9.24-SNAPSHOT` as the CARDS version as that is what `cards/cards:latest` is when `setup.sh` was ran but then `cards/cards:latest` is replaced with a new image of the generic CARDS platform but this time of version `0.9.25-SNAPSHOT` this build should fail.

sdumitriu commented on setup.sh:
```diff
@@ -1,45 +1,110 @@
 
 # Remove the cards4 prefix to get the short unprefixed name
 PROJECT_SHORTNAME=${PROJECT_CODENAME#cards4}
+if [[ $? == 1 ]]
+then
+  exit 1
+fi
```
```suggestion
```

sdumitriu committed:
Exit the setup script whenever the user Cancels a dialog

Co-authored-by: Michael Lescisin <IntegralProgrammer@users.noreply.github.com>

sdumitriu committed:
Check if the local cards image has the expected version

sdumitriu committed:
Bugfix: local snapshot version fails to be detected

sdumitriu committed:
Also check that the image exists locally if we build with the `local` profile

sdumitriu committed:
Fix build warning due to unknown maven plugin config

IntegralProgrammer commented on pom.xml:
```diff
@@ -31,9 +31,12 @@
-  <properties>
-    <cards.version>$CARDS_VERSION$</cards.version>
-  </properties>
+  <scm>
+    <connection>scm:git:git://github.com/data-team-uhn/$PROJECT_SHORTNAME$.git</connection>
+    <developerConnection>scm:git:git@github.com:data-team-uhn/$PROJECT_SHORTNAME$.git</developerConnection>
+    <url>https://github.com/data-team-uhn/$PROJECT_SHORTNAME$/tree/master/</url>
+    <tag>HEAD</tag>
+  </scm>
```
This assumes that the project's source code will always be located under `github.com/data-team-uhn/` and will use a _default_ branch name of `master`.

How about, during the execution of `setup.sh`, we prompt for the project repository location and default branch name, reconfigure the local git repository to set the `origin` remote to the given remote repository location and set the local branch to the provided _default_ branch name. These settings would then be reflected in this `<scm>` part of this `pom.xml` file?

sdumitriu replied:
The way this is meant to be used is:
- create a new project in github, picking the template repo as a template
- clone locally
- run the setup script
So we should already have a github repo set up when running the script. The default branch name should be different. We should update the pom to reflect the actual origin URL, and we could use the repo name in the script as an initial prompt for the `PROJECT_SHORTNAME` value.

IntegralProgrammer replied:
How about this:

- When `setup.sh` runs it will use `git remote get-url origin` to get the URL of the recently cloned git repository for the new project (as opposed to its template).
- Based on the result of `git remote get-url origin` we can provide a suggestion for the `PROJECT_SHORTNAME` value.
- The default branch name can be set by the `setup.sh` script by having it call `git rev-parse --abbrev-ref HEAD`. This way, the default branch name of this template repository will match with the default branch name of the new project repository.

sdumitriu replied:
Precisely my plan!

sdumitriu committed:
Use the actual repo URL in the pom file, and the repo name as the suggested project codename

IntegralProgrammer commented on setup.sh:
```diff
@@ -16,6 +16,20 @@
 
+# Check the repo URL and main branch
+REPO_URL=$(git remote get-url origin)
+if [[ $REPO_URL == *data-team-uhn/cards-project-template.git ]]
```
This check should be fixed as, if the repository was cloned with `git clone https://github.com/data-team-uhn/cards-project-template`, `git remote get-url origin` will return `https://github.com/data-team-uhn/cards-project-template` (without the trailing `.git`) and thus the `setup.sh` script will proceed even though the _git remote_ is set to the _template_ repository as opposed to the new _project_ repository.

I think checking if `$REPO_URL` ends with `/cards-project-template` or `/cards-project-template.git` would be sufficient. Including the `data-team-uhn` part of the URL would cause this check to break (always pass) on forks of this repository.

sdumitriu replied:
Done.

IntegralProgrammer commented on setup.sh:
```diff
@@ -16,6 +16,20 @@
+fi
+# github uses two different formats for the repo url, ssh or http, while Maven expects a different format.
+# Convert the git remote URL to the format supported by Maven
+REPO_NAME=$(echo -n $REPO_URL | sed -E -e 's/.*data-team-uhn[/:](.*).git/\1/')
```
Similar to my above comment, this line should be amended to handle the case where `$REPO_URL` does not contain the trailing `.git` or the `data-team-uhn` organization.

sdumitriu replied:
Done.

IntegralProgrammer commented on setup.sh:
```diff
@@ -16,6 +16,20 @@
+# github uses two different formats for the repo url, ssh or http, while Maven expects a different format.
+# Convert the git remote URL to the format supported by Maven
+REPO_NAME=$(echo -n $REPO_URL | sed -E -e 's/.*data-team-uhn[/:](.*).git/\1/')
+MVN_READ_URL=$(echo -n $REPO_URL | sed -E -e 's/(git@|https:\/\/)/scm:git:git:\/\//' | sed -E -e 's/github.com:/github.com\//' | sed -e 's/\//\\\//g')
+MVN_WRITE_URL=$(echo -n $REPO_URL | sed -E -e 's/(git@|https:\/\/)/scm:git:git@/' | sed -E -e 's/github.com\//github.com:/' | sed -e 's/\//\\\//g')
+MVN_URL=$(echo -n $REPO_URL | sed -E -e 's/(git@|https:\/\/)/https:\/\//' | sed -E -e 's/github.com:/github.com\//' | sed -E -e "s/.git$/\/tree\/$(git rev-parse --abbrev-ref HEAD)\//" | sed -e 's/\//\\\//g')
```
Can this be changed to handle the case where the git repository for the project is not hosted on GitHub?

sdumitriu replied:
This isn't a concern right now, I'd rather keep it simple.

sdumitriu committed:
Make sure .git is always in the repo URL

sdumitriu committed:
Allow forks of the template repo to also be detected as templates

IntegralProgrammer commented on setup.sh:
```diff
@@ -58,90 +141,85 @@ else
+# Ask about the permission scheme
 
-  if [[ $HAS_BACKEND != "yes" && $HAS_BACKEND != "no" ]]
+# Get the list of features already included in the base distribution
+# Then get the list of all features known, excluding the cards4* projects
+# Then subtract included features from the list of all known features
```
Does this comment belong here? It's the `permissionlist` list that is being built as opposed to the `featurelist` which is built later in this script and described by the same comment https://github.com/data-team-uhn/cards-project-template/pull/1/files#diff-4209d788ad32c40cbda3c66b3de47eefb929308ca703bb77a6382625986add17R177-R179

sdumitriu replied:
It used to be correct, but it doesn't reflect the current code.

sdumitriu replied:
Removed.

sdumitriu committed:
Remove wrong comment

IntegralProgrammer reviewed (approved):
Looks good to me. I was able to create a test project based on the released CARDS-0.9.24.1 and was able to start a production image of it using Docker Compose as well as run it with `start_cards.sh`. Any other potential improvements can be done in other PRs. :white_check_mark:
